[MachineOutliner] Preserve regmasks in calls to outlined functions

nocchijiang opened the pull request:
When emitting calls to an outlined function, the register masks from the outlined sequence are lost. The AArch64CollectLOH pass, which I plan to move to PreEmitPass2 (positioned after MachineOutliner), relies on accurate register masks. This patch ensures that regmasks are correctly preserved in the outlined calls, maintaining the required accuracy for subsequent passes.

nocchijiang committed:
[MachineOutliner] Preserve regmasks in calls to outlined functions

When emitting calls to an outlined function, the register masks from the
outlined sequence are lost. The AArch64CollectLOH pass, which I plan to
move to PreEmitPass2 (positioned after MachineOutliner), relies on
accurate register masks. This patch ensures that regmasks are correctly
preserved in the outlined calls, maintaining the required accuracy for
subsequent passes.

kyulee-com commented on llvm/test/CodeGen/AArch64/machine-outliner-regmask.mir:
```diff
@@ -0,0 +1,94 @@
+# RUN: llc -mtriple=aarch64-apple-ios -run-pass=prologepilog -run-pass=machine-outliner %s -o - | FileCheck %s
+--- |
+  declare swiftcc void @bar()
+  declare void @baz(i32, i32, i32) #0
```
Replace all `#0` by `minsize` or just drop it as there is no def.

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -1130,6 +1131,12 @@ bool MachineOutliner::outline(
           SmallSet<Register, 2> InstrUseRegs;
           for (MachineOperand &MOP : MI->operands()) {
+            // Collect all regmasks. Merge them in the end.
+            if (MOP.isRegMask()) {
```
I see you attempt to optimize the case with 1 just for copying the pointer. But I'd simply merging reg mask in place here, without using the extra set. 
Basically, you could declare `RegMaskSize` outside the loop, as it's a constant for the function.
```
uint32_t *RegMask = nullptr;
...
            if (MOP.isRegMask()) {
              const uint32_t *CurrRegMask = MOP.getRegMask();
              if (!RegMask) {
                RegMask = MF->allocateRegMask();
                memcpy(RegMask, CurrRegMask, sizeof(RegMask[0]) * RegMaskSize);
              } else {
                for (unsigned I = 0; I < RegMaskSize; ++I)
                  RegMask[I] &= CurrRegMask[I];
              }
              continue;
            }
...
            if (RegMask)
              CallInst->addOperand(MachineOperand::CreateRegMask(RegMask));

  ```

nocchijiang replied:
I believe 1-regmask is the most common case and the memory allocation in `MF->allocateRegMask()` should be avoided if possible.

ellishg replied:
The size of `RegMasks` depends on the number of call instructions in the outlined candidate, and I believe this usually one. However, I think you might be forgetting that non-call instruction can clobber a register. My understanding of `RegMask` is that it tells you which registers could be changed after executing that instruction.

https://github.com/llvm/llvm-project/blob/76dba2ea02048b587d10ccdf777b0875369f362c/llvm/include/llvm/CodeGen/MachineOperand.h#L921-L927

But an outlined function might have other instructions as well as call instructions. I think you need to mark `MOP.isDef()` registers as clobbered in the mask too (I believe this is what @MatzeB was alluding to).

In that case, we will likely need to copy the mask using `allocateRegMask()` anyway, so I think @kyulee-com suggestion might be worth considering.

nocchijiang replied:
If we need to mark registers from def operands in the mask, I have a follow-up question: should we drop all the def operands in favor of using regmask, or should we add the masked registers to the def list as well? My understanding is that regmask and def operands serve different purposes - regmask is primarily used to describe calling conventions, while def operands provide additional information specific to the callee. Including all defs in the regmask seems redundant to me. Could anyone clarify the intended distinction between these two, or share best practices for handling this case?

nocchijiang committed:
add missing attribute

kyulee-com reviewed (approved):
LGTM. But wait for a few days to get reviews, if any, from others.

ellishg commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -1153,6 +1160,24 @@ bool MachineOutliner::outline(
+                MF->getSubtarget().getRegisterInfo()->getNumRegs();
+            unsigned Size = MachineOperand::getRegMaskSize(NumRegs);
+            memset(RegMask, UINT32_MAX, Size * sizeof(RegMask[0]));
+            for (const uint32_t *Mask : RegMasks) {
+              for (unsigned I = 0; I < Size; ++I)
+                RegMask[I] &= Mask[I];
+            }
```
```suggestion
            for (const uint32_t *Mask : RegMasks)
              for (unsigned I = 0; I < Size; ++I)
                RegMask[I] &= Mask[I];
```

nocchijiang committed:
remove redundant braces

Co-authored-by: Ellis Hoag <ellis.sparky.hoag@gmail.com>

kyulee-com commented:
This seems a good fix. Any other comments? @aemerson @ornata @MatzeB

aemerson commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -1153,6 +1160,23 @@ bool MachineOutliner::outline(
+            unsigned NumRegs =
+                MF->getSubtarget().getRegisterInfo()->getNumRegs();
+            unsigned Size = MachineOperand::getRegMaskSize(NumRegs);
+            memset(RegMask, UINT32_MAX, Size * sizeof(RegMask[0]));
```
I feel like this is too low level of an abstraction to be using to construct a new regmask. Can you move this into MachineFunction or some other appropriate place to provide a cleaner API?

nocchijiang replied:
These lines are copied from `MachineFunction::allocateRegMask`. Therefore, the most straightforward approach is to extend it while maintaining its behavior unchanged for existing users with default parameter values. If you have any better ideas, please let me know.

MatzeB replied:
This is also misleading. The 2nd parameter of `memset` is NOT an UINT32_MAX but is an `int` and interpreted as a byte...

kyulee-com commented on llvm/include/llvm/CodeGen/MachineFunction.h:
```diff
@@ -1119,8 +1119,10 @@ class LLVM_ABI MachineFunction {
+  /// Allocate and initialize a register mask with all elements set to
+  /// \p InitValue and size reported via \p OutSize if nonnull.
+  uint32_t *allocateRegMask(uint32_t InitValue = 0,
+                            unsigned *OutSize = nullptr);
```
Can we create another API like the one below and use it to redefine this for backward compatibility?
```
MutableArrayRef<uint32_t> allocateRegMaskArray(uint8_t InitValue = 0);
```

Additionally, looking at the usage of `allocateRegMask()`, there are many cases where it is used for cloning a RegMask. I think creating another API, as shown below, could utilize the above `allocateRegMaskArray()` to simplify the existing code. However, this might be better for a separate PR.
```
uint32_t *cloneRegMask(const uint32_t *RegMask);
```

aemerson replied:
Agreed, it's a step in the right direction but I think we can do better.

MatzeB replied:
Please don't change the existing `allocatRegMask()` function for this it doesn't need and shouldn't have it IMO!

MatzeB commented:
Not sure I have the full context here, but this feels like it is attempting to construct some clobber mask for an outlined call by merging all the register masks of the calls in the newly outlined function. That seems wrong to me as it completely ignores any other register usage (normal defs can clobber registers just as well as calls with clobbers can). So this seems incomplete/wrong to me as-is...

nocchijiang commented:
I think it's necessary to pause and reach an agreement before proceeding with further changes.

To answer @MatzeB 's question first: I apologize for not posting the link to the RFC earlier: https://discourse.llvm.org/t/loh-conflicting-with-machineoutliner/83279. Here’s a summary: I plan to move AArch64CollectLOH to addPreEmitPass2 to allow the outliner to optimize more. Historically AArch64CollectLOH has made certain assumptions about its input and has accepted some trade-offs for compile-time performance (some of which I believe is actually written by @MatzeB). These will be invalidated with the proposed change.

What this PR addresses falls in the latter category: AArch64CollectLOH does not perform accurate register liveness analysis but only looks into regmasks and defs to determine if a given register is clobbered. MachineOutliner emits calls to outlined sequences, which probably contains calls to another function which clobbers caller-saved registers. This is not reflected in the defs but in the regmask. Therefore, the accuracy of regmask is key to accurate LOH emissions, which MachineOutliner lacks right now.

Regarding concerns about the `allocateRegMask` refactoring: regmask has been seriously lacking proper abstraction and encapsulation. A simple global search on `allocateRegMask` reveals a lot of instances of implementation detail leaks. We need to clearly define the aspects we want to improve in this PR; otherwise, it could become endless and extend far beyond its scope. `MutableArrayRef<uint32_t>` suggested by @kyulee-com indeed avoids explicit size reporting but offers no improvement in other aspects (initial value setting, cloning, manipulation, etc.).

kyulee-com commented:
The effective set of Def and Use registers for the given range (the outlined instruction sequence) has already been computed by reverse traversing the instructions. Here, additionally it merges the reg masks that point to clobbering registers in call instructions. This ensures that all definitions are covered either in the Def registers or these combined reg masks. Conservatively, we could also set these bits for the Def registers in these masks, but I don't think it's necessary as we already have the Def registers mentioned above. @MatzeB Do you think the reg masks should have all Def register bits regardless?

It's worth nothing we need custom reg masks because the outlined function call does not follow a typical calling convention; instead, it's a naked call for the range of instructions without forming a frame. So, we should conservatively merge all clobbered register effects into the masks, which this PR seems to fix.

Regarding my suggestion, `MutableArrayRef<uint32_t> allocateRegMaskArray(uint8_t InitValue = 0);`, I didn't intend to modify the API of the existing `uint32_t *allocateRegMask();`. Instead, it can be implemented in a way that preserves the existing behavior while also providing the array size as needed in this context here. @MatzeB @aemerson What do you suggest here?
```
uint32_t *allocateRegMask() { return allocateRegMaskArray().data(); }
```

nocchijiang committed:
encapsulate regmask creation

nocchijiang commented:
Having been occupied with the CNY and other company-related matters, I have finally had the opportunity to work on this PR. Given that @MatzeB has not provided any further comments, I presume you have been familiar with the context. I have addressed all comments from the participating reviewers in the latest commit.

nocchijiang commented:
@aemerson @MatzeB ping

ellishg reviewed (changes requested):
As described in this comment: https://discourse.llvm.org/t/loh-conflicting-with-machineoutliner/83279/12, I don't think it's correct to move LOH after the MachineOutliner. Therefore this PR is not needed.
